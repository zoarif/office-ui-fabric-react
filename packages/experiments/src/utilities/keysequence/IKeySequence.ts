import { ktpPrefix, ktpSeparator } from '../keytip/KeytipUtils';
import { find } from '../../Utilities';

export interface IKeySequence {
  keys: string[];
}

/**
 * Tests for equality between two IKeySequences
 *
 * @param seq1 - First IKeySequence
 * @param seq2 - Second IKeySequence
 * @returns {boolean} T/F if the two sequences are equal
 */
export function keySequencesAreEqual(seq1: IKeySequence, seq2: IKeySequence): boolean {
  const keyCodes1 = seq1.keys.join();
  const keyCodes2 = seq2.keys.join();
  return keyCodes1 === keyCodes2;
}

/**
 * Tests for equality between two arrays of IKeySequence
 *
 * @param seq1 - First IKeySequence[]
 * @param seq2 - Second IKeySequence[]
 * @returns {boolean} T/F if the sequences are equal
 */
export function fullKeySequencesAreEqual(seq1: IKeySequence[], seq2: IKeySequence[]): boolean {
  if (seq1.length !== seq2.length) {
    return false;
  }

  for (let i = 0; i < seq1.length; i++) {
    if (!keySequencesAreEqual(seq1[i], seq2[i])) {
      return false;
    }
  }

  return true;
}

/**
 * Tests if 'seq' is present in 'sequences'
 *
 * @param sequences - Array of IKeySequence
 * @param seq - IKeySequence to check for in 'sequences'
 * @returns {boolean} T/F if 'sequences' contains 'seq'
 */
export function keySequencesContain(sequences: IKeySequence[], seq: IKeySequence): boolean {
  return !!find(sequences, (sequence: IKeySequence) => {
    return keySequencesAreEqual(sequence, seq);
  });
}

/**
 * Method returns true if the key squence of the object with minimum length is in the other key sequence.
 * If the minium length is zero, then it will default to false.
 *
 * @param seq1 - First IKeySequence
 * @param seq2 - Second IKeySequence
 * @returns {boolean} T/F if seq1 starts with seq2, or vice versa
 */
export function keySequenceStartsWith(seq1: IKeySequence, seq2: IKeySequence): boolean {
  const keyCodes1 = seq1.keys.join();
  const keyCodes2 = seq2.keys.join();
  if (keyCodes1.length === 0 || keyCodes2.length === 0) {
    return false;
  }
  return keyCodes1.indexOf(keyCodes2) === 0 || keyCodes2.indexOf(keyCodes1) === 0;
}

/**
 * Converts a whole set of KeySequences into one keytip ID, which will be the ID for the last keytip sequence specified
 * keySequences should not include the initial keytip 'start' sequence
 *
 * @param keySequences - Full path of IKeySequences for one keytip
 * @returns {string} String to use for the keytip ID
 */
export function convertSequencesToKeytipID(keySequences: IKeySequence[]): string {
  return keySequences.reduce((prevValue: string, keySequence: IKeySequence): string => {
    return prevValue + ktpSeparator + keySequence.keys.join(ktpSeparator);
  }, ktpPrefix);
}