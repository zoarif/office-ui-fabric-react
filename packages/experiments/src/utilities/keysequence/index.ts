export * from './IKeySequence';
export * from './IKeytipTransitionKey';