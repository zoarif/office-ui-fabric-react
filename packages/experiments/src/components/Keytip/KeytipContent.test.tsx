import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { IKeySequence } from '../../utilities/keysequence/IKeySequence';
import { KeytipContent } from './KeytipContent';

const sequence: IKeySequence[] = [{ keys: ['a'] }];
const keyCont = 'A';

describe('Keytip', () => {
  it('renders visible Keytip correctly', () => {
    const componentContent = renderer.create(<KeytipContent visible={ true } content={ keyCont } keySequences={ sequence } />);
    const tree = componentContent.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders visible disabled Keytip correctly', () => {
    const componentContent = renderer.create(
      <KeytipContent
        visible={ true }
        disabled={ true }
        content={ keyCont }
        keySequences={ sequence }
      />);
    const tree = componentContent.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders invisible Keytip correctly', () => {
    const componentContent = renderer.create(<KeytipContent visible={ false } content={ keyCont } keySequences={ sequence } />);
    const tree = componentContent.toJSON();
    expect(tree).toMatchSnapshot();
  });
});