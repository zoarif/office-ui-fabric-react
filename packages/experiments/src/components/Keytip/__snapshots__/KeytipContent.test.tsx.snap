// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Keytip renders invisible Keytip correctly 1`] = `
<div
  className=

      {
        background-color: #212121;
        visibility: hidden;
      }
>
  <span
    className=

        {
          background-color: #212121;
          color: #eaeaea;
          display: inline-block;
          height: 17px;
          line-height: 17px;
          min-width: 11px;
          padding-left: 3px;
          padding-right: 3px;
          text-align: center;
        }
    id="ktp-a"
  >
    A
  </span>
</div>
`;

exports[`Keytip renders visible Keytip correctly 1`] = `
<div
  className=

      {
        background-color: #212121;
      }
>
  <span
    className=

        {
          background-color: #212121;
          color: #eaeaea;
          display: inline-block;
          height: 17px;
          line-height: 17px;
          min-width: 11px;
          padding-left: 3px;
          padding-right: 3px;
          text-align: center;
        }
    id="ktp-a"
  >
    A
  </span>
</div>
`;

exports[`Keytip renders visible disabled Keytip correctly 1`] = `
<div
  className=

      {
        background-color: #212121;
        opacity: 0.5;
      }
>
  <span
    className=

        {
          background-color: #212121;
          color: #b1b1b1;
          display: inline-block;
          height: 17px;
          line-height: 17px;
          min-width: 11px;
          padding-left: 3px;
          padding-right: 3px;
          text-align: center;
        }
    id="ktp-a"
  >
    A
  </span>
</div>
`;
